### PR TITLE
[HWASAN] Update dbg.assign intrinsics in HWAsan pass

### LINKED ISSUE #76545: HWASAN: No DW_OP_LLVM_tag_offset with opt above -O0
DW_OP_LLVM_tag_offset is present in -O0, but not -O1 https://godbolt.org/z/3dj6bzz4z

Reasons is `AssignmentTrackingPass` and `-instcombine-lower-dbg-declare` replace `@llvm.dbg.declare` with `@llvm.dbg.assign` but we don't attach DW_OP_LLVM_tag_offset to `@llvm.dbg.assign`.

### PR DESCRIPTION
**Pull request by OCHyams**
llvm.dbg.assign intrinsics have 2 {value, expression} pairs; fix hwasan to
update the second expression.

Fixes #76545. This is #78606 rebased and with the addition of DPValue handling.
Note the addition of --try-experimental-debuginfo-iterators in the tests and
some shuffling of code in MemoryTaggingSupport.cpp.

I opened a new PR because I can't find a way to change the status of that one
from merged.

This depends on #79816 (which is included as a patch in this pull request and
should be ignored - in llvm/lib/IR/DebugInfoMetadata.cpp).

---

Given the other version of the patch was in and out a few times, I thought it worth mentioning that a self-host build of clang with `-O2 -g -fsanitzer=HWAddress` completes without issue with these patches.

### CODE REVIEW

**Review by vitalybuka — APPROVED** [review]

File: llvm/lib/Transforms/Utils/MemoryTaggingSupport.cpp [hunk 1]
```diff
@@ -110,6 +110,24 @@ Instruction *getUntagLocationIfFunctionExit(Instruction &Inst) {
...
+      if (auto *AI = dyn_cast_or_null<AllocaInst>(V)) {
+        if (!isInterestingAlloca(*AI))
+          return;
+        AllocaInfo &AInfo = Info.AllocasToInstrument[AI];
+        auto &DPVVec = AInfo.DbgVariableRecords;
+        if (DPVVec.empty() || DPVVec.back() != &DPV)
+          DPVVec.push_back(&DPV);
+      }
+    };
+    for (Value *V : DPV.location_ops())
```
Comment (fmayer) [inline llvm/lib/Transforms/Utils/MemoryTaggingSupport.cpp:125]: optional nit: for_each

File: llvm/lib/Transforms/Utils/MemoryTaggingSupport.cpp [hunk 2]
```diff
@@ -138,30 +156,20 @@ void StackInfoBuilder::visit(Instruction &Inst) {
...
-        if (!isInterestingAlloca(*AI))
-          continue;
-        AllocaInfo &AInfo = Info.AllocasToInstrument[AI];
-        auto &DPVVec = AInfo.DbgVariableRecords;
-        if (DPVVec.empty() || DPVVec.back() != &DPV)
-          DPVVec.push_back(&DPV);
-      }
-    }
+    };
+    for (Value *V : DVI->location_ops())
```
Comment (fmayer) [inline llvm/lib/Transforms/Utils/MemoryTaggingSupport.cpp:169]: optional nit: for_each

File: llvm/test/CodeGen/AArch64/dbg-assign-tag-offset.ll [hunk 3]
```diff
@@ -0,0 +1,72 @@
+; RUN: llc -filetype=obj -o - %s | llvm-dwarfdump - | FileCheck %s
+; RUN: llc --try-experimental-debuginfo-iterators -filetype=obj -o - %s | llvm-dwarfdump - | FileCheck %s
+
```
Comment (fmayer) [inline llvm/test/CodeGen/AArch64/dbg-assign-tag-offset.ll:3]: Please add a sentence what we are testing here.

**Comment by fmayer** [comment]
> HWASAN is LGTM @eugenis and @pcc for Mem tagging

This doesn't actually change MTE. MemoryTaggingSupport is used by MTE stack tagging and HWASan. For MTE,  I _think_ we need something similar to 

```
      DVI->replaceVariableLocationOp(OldAI, Info.AI);
```

in the Aarch64MemoryTagging pass. But that's for another CL.

**Comment by OCHyams** [comment]
@fmayer are you happy with the changes?

**Review by fmayer — APPROVED** [review]

**Pull request merged**